packages/ai-jsx/src/core/conversation.tsx

```diff
@@ -309,3 +323,204 @@ export async function* ShowConversation(
+      if (leastImportantDescendant !== undefined && compareImportance(leastImportantDescendant, current) < 0) {
+        current = leastImportantDescendant;
+      }
+    });
+
+    return current;
+  }
+
+  function aggregateCost(roots: TreeNode[]): number {
+    return roots.reduce((cost, node) => cost + node.cost, 0);
+  }
+
+  /** Replaces a single ShrinkableTreeNode in the tree. */
+  async function replaceNode(roots: TreeNode[], nodeToReplace: ShrinkableTreeNode): Promise<TreeNode[]> {
+    const newRoots = await Promise.all(
+      roots.map<Promise<TreeNode[]>>(async (root) => {
+        if (root === nodeToReplace) {
+          return conversationToTreeRoots(root.element.props.replacement);
+        }
+
+        if (root.type !== 'shrinkable') {
+          return [root];
+        }
+
+        // Look for a replacement among the children and recalculate the cost.
+        const replacementChildren = await replaceNode(root.children, nodeToReplace);
+        return [
+          {
+            type: 'shrinkable',
+            element: root.element,
+            cost: aggregateCost(replacementChildren),
+            children: replacementChildren,
+          },
+        ];
+      })
+    );
+
+    return newRoots.flat(1);
+  }
+
+  /** Converts the shrinkable tree into a single AI.Node for rendering. */
+  function treeRootsToNode(roots: TreeNode[]): AI.Node {
+    return roots.map((root) => (root.type === 'immutable' ? root.element : treeRootsToNode(root.children)));
+  }
+
+  const memoized = memo(children);
+
+  // If there are no shrinkable elements, there's no need to evaluate the cost.
+  const shrinkableOrConversationElements = (
+    await render(memoized, {
+      stop: (e) => isConversationalComponent(e) || e.tag === InternalShrinkable,
+    })
+  ).filter(AI.isElement);
+  if (!shrinkableOrConversationElements.find((value) => value.tag === InternalShrinkable)) {
+    return shrinkableOrConversationElements;
+  }
+
+  let roots = await conversationToTreeRoots(shrinkableOrConversationElements);
+  while (aggregateCost(roots) > budget) {
+    const nodeToReplace = leastImportantNode(roots);
+    if (nodeToReplace === undefined) {
+      // Nothing left to replace.
+      break;
+    }
+
+    logger.debug(
+      {
+        node: debug(nodeToReplace.element.props.children, true),
+        importance: nodeToReplace.element.props.importance,
+        replacement: debug(nodeToReplace.element.props.replacement, true),
+        nodeCost: nodeToReplace.cost,
+        totalCost: aggregateCost(roots),
+        budget,
+      },
+      'Replacing shrinkable content'
+    );
+
+    // N.B. This currently quadratic in that each time we replace a node we search the entire
+    // tree for the least important node, and then search _again_ to replace it. If we end up
+    // doing many replacements we should be smarter about this.
+    roots = await replaceNode(roots, nodeToReplace);
```

Comment:
Good note, but in practice, I think the "we end up doing many replacements" is unlikely.